domain/application/state/resource.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+
+	coreresource "github.com/juju/juju/core/resource"
+	"github.com/juju/juju/domain/application"
+	"github.com/juju/juju/domain/application/charm"
+	apperrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// buildResourceToAdd creates resources to add based on provided app and charm
+// resources.
+// Returns a slice of resourceToAdd and an error if any issues occur during
+// creation. Validates app resources against charm resources, warning if
+// mismatches are found.
+// Returns an error satisfying [apperrors.InvalidResourceArgs] if there is
+// unexpected resource in argument that are not defined in charm.
+func (st *State) buildResourceToAdd(appUUID, charmUUID string,
+	appResources []application.AddApplicationResourceArg,
+	charmResources map[string]charm.Resource) (
+	[]resourceToAdd, error) {
+	resources := make([]resourceToAdd, 0, len(appResources))
+	appResourceSet := set.NewStrings()
+	charmResourceSet := set.NewStrings()
+	now := st.clock.Now()
+	for _, r := range appResources {
+		appResourceSet.Add(r.Name)
+		uuid, err := coreresource.NewUUID()
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		resources = append(resources, resourceToAdd{
+			UUID:            uuid.String(),
+			ApplicationUUID: appUUID,
+			CharmUUID:       charmUUID,
+			Name:            r.Name,
+			Revision:        r.Revision,
+			Origin:          r.Origin.String(),
+			State:           coreresource.StateAvailable.String(),
+			CreatedAt:       now,
+		})
+	}
+	for _, res := range charmResources {
+		charmResourceSet.Add(res.Name)
+	}
+
+	unexpectedResources := appResourceSet.Difference(charmResourceSet)
+	missingResources := charmResourceSet.Difference(appResourceSet)
```

Comment:
If either of these two sets are not empty. It's a client error which should be raised. As the check is being added, this is the best time to create the error.